feat(group_theory/iwasawa.lean): add a proof of the Iwasawa criterion for simplicity

AntoineChambert-Loir opened the pull request:
Add the Iwasawa criterion for simplicity.
This criterion can be used in the proof of simplicity of some classical
groups such as the alternating group or the projective special linear group
(these applications are TODO).



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Commit:
feat(group_theory/iwasawa.lean): add a proof of the Iwasawa criterion for simplicity

Add the Iwasawa criterion for simplicity.
This criterion can be used in the proof of simplicity of some classical
groups such as the alternating group or the projective special linear group
(these applications are TODO).

Commit:
style: remove the final #lint

jcommelin commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+
+**Lemma.** If the action is 2-transitive, then it is primitive.
+
+**Theorem (Iwasawa criterion).** Let G be a group with a primitive action on a set X.
```
Thanks for this module docstring at the top of the file!
Could you please cross-link to the Lean code below, by adding the name of Lean theorems/defs (in backticks) to this list of thms/defs? The generated html documentation will automatically turn those into hyperlinks and people who open this file will quickly know the Lean name of the main result.

jcommelin commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+/-- A 2-transitive action is primitive -/
+/- Part of the proof establishes that stabilizers of n-transitive
+actions are (n-1)-transitive. Todo: rewrite using this. -/
+theorem is_primitive_of_two_trans 
```
It seems that you are using this notion of primitive actions quite a bit below. I think it makes sense to create a new definition for this, and use it everywhere.

jcommelin replied:
Apologies. The above comment was apparently written when I was not completely awake.

jcommelin commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+
+/-- An subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+class subgroup.is_maximal (K : subgroup G) : Prop := 
+  (out : is_coatom K)  
```
```suggestion
(out : is_coatom K)  
```

jcommelin commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+theorem is_maximal_iff {K: subgroup G} : K.is_maximal ↔
+  K ≠ ⊤ ∧ ∀ (H: subgroup G) g, K ≤ H → g ∉ K → g ∈ H → H = ⊤ :=
+begin
+  split, 
```
Please put `{ }` around the subproofs that work on the two goals.
I see that you are structuring your proof with newlines, which also indicates the subproofs, but the mathlib style is to use `{ }`-blocks around subproofs.

jcommelin commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+  have zKH : K ≠ H := 
+  begin 
+    simp, intro z, rw z at hgK, exact hgK hgH,
+  end,
```
Two remarks:

1. You can save two lines, by just writing `{ some, tactic, proof },` instead of the `begin .. end` block.
2. Please replace `simp` by the output of `squeeze_simp`. (In VScode, you can even click on the output in the goal view, and it will automatically replace it.) The reason we don't want a bare `simp` halfway a proof is the following: if someone adds new simp-lemmas to file `X`, and in a later PR file `X` is added as import to one of the dependencies of this Iwasawa file, all of a sudden, this `simp` might do more/different stuff from what it is doing now. And then this proof breaks. But the person who is adding this import maybe doesn't understand the maths in this file. So now they have a PR with a broken proof in random part of the library that they don't know how to fix.
Long story short: `squeeze_simp` helps a lot with long-term maintenance of proofs. See also https://leanprover-community.github.io/glossary.html#non-terminal-simp

AntoineChambert-Loir replied:
Thanks! The main reason for my `begin … end` blocks is that I wrote  `theorem …… := …`,
and the `{… }` syntax didn't work until I changed the `:=` to `,`.

Regarding 2, this is an issue I had imagined you could have, when people rewrite/extend tactics.
Most often, `squeeze_simp` suggests a `simp only [do_stuff]` 
and I see that `rw do_stuff` suffices.
Is it enough to leave the `simp only` or is it better to change it to `rw`?

jcommelin replied:
Usually `simp only` is fine. `rw` is slightly faster, but is a bit less powerful than `simp only`. I wouldn't worry too much for now. `simp only` is stable and pretty fast.

jcommelin commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+end
+
+/-- f(f⁻¹(t)) = f.range ∩ t  -/
+lemma map_comap_of {α β : Type*} (f: α → β) (t: set β):
```
I think the more predictable name would be to use `image_preimage_eq`. Also, these lemmas should move to some basic file about `set`.

jcommelin commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+  by rw [← set.image_univ, ← set.image_inter_preimage f set.univ t, set.univ_inter]
+
+/-- If f is surjective, then f(f⁻¹(t)) = t -/
+lemma map_comap_of_surjective {α β : Type*} (f : α → β) (t : set β) :
```
Idem.

AntoineChambert-Loir replied:
How to guess whether they exist in some form ?

jcommelin replied:
You can try to prove them `by library_search`. That might tell you they already exist.

tb65536 commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+    simp, exact hn,
+  end,
+  apply is_faithful.eq_of_smul_eq_smul, exact N_triv, exact hN N_bot,
+end
```
```suggestion
theorem is_simple (is_nontrivial : nontrivial G) (is_perfect : commutator G = ⊤)
  (is_prim : is_primitive G X) (is_faithful : has_faithful_scalar G X)
  (IwaS : has_iwasawa_structure G X) : is_simple_group G :=
begin
  refine ⟨is_nontrivial.exists_pair_ne, λ N nN, _⟩,
  cases (or_iff_not_imp_left.mpr (Iwasawa_mk G X is_prim IwaS N nN)),
  { refine or.inl (N.eq_bot_iff_forall.mpr (λ n hn, _)),
    apply is_faithful.eq_of_smul_eq_smul,
    intro x,
    rw one_smul,
    exact set.eq_univ_iff_forall.mp h x ⟨n, hn⟩ },
  { exact or.inr (top_le_iff.mp (le_trans (ge_of_eq is_perfect) h)) },
end
```

tb65536 commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,655 @@
+-/
+
+exact contains_commutators_of (N:subgroup G) nN (IwaS.T x) 
+  (is_top_of_NTx x) (IwaS.is_comm x),
```
```suggestion
  intros N nN hNX,
  haveI := prim_to_transitive G X is_prim N nN hNX,
  obtain x := is_prim.is_nonempty.some,
  refine contains_commutators_of N nN (IwaS.T x) _ (IwaS.is_comm x),
  rw [eq_top_iff, ←IwaS.is_generator, supr_le_iff],
  intro y,
  obtain ⟨g, rfl⟩ := mul_action.exists_smul_eq N x y,
  rw [subgroup.smul_def, IwaS.is_conj g x],
  rintros _ ⟨k, hk, rfl⟩,
  have hg' : ↑g ∈ N ⊔ IwaS.T x := subgroup.mem_sup_left (subtype.mem g),
  have hk' : k ∈ N ⊔ IwaS.T x := subgroup.mem_sup_right hk,
  exact (N ⊔ IwaS.T x).mul_mem ((N ⊔ IwaS.T x).mul_mem hg' hk') ((N ⊔ IwaS.T x).inv_mem hg'),
```

AntoineChambert-Loir replied:
Cool to see this so much shorter ! 
But isn't it sad that the proofs lose all verbose comments that (possibly) make it easier to read by a human being ?

tb65536 replied:
You can add comments back in, if you want.

tb65536 reviewed (commented):
Golfed the last two proofs.

Commit:
style (group_theory/iwasawa.lean): shorten proofs and remove unused stuff

Use shortened proofs as suggested by @tb65536
Use set.image_preimage and remove unused basic lemmas
Use nontrivial and remove useless intermediate results

Commit:
style(group_theory/iwasawa.lean): add copyright, remove #print

Commit:
style(group_theory/iwasawa.lean): put docstring first

tb65536 commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,487 @@
+        /- on veut applique x sur y
+          mais il faut un élément x' ≠ x à appliquer sur un élément y' ≠ y    -/
+        obtain ⟨x', hx'⟩ := @exists_ne _ hX x,
+        obtain ⟨y', hy'⟩ := @exists_ne _ hX y,
```
Does this work?
```suggestion
        obtain ⟨x', hx'⟩ := exists_ne x,
        obtain ⟨y', hy'⟩ := exists_ne y,
```

AntoineChambert-Loir replied:
No ! I get the following error message : 

```
failed to synthesize type class instance for
G : Type u_1,
_inst_1 : group G,
X : Type u_2,
_inst_2 : mul_action G X,
h2X :
  ∀ (x₁ x₂ y₁ y₂ : X),
    x₁ ≠ x₂ → y₁ ≠ y₂ → (∃ (g : G), g • x₁ = y₁ ∧ g • x₂ = y₂),
hX : nontrivial X,
x y : X
⊢ nontrivial X
```

which I interpret as Lean telling me it didn't find by itself that it could use `hX`.

tb65536 replied:
The issue seems to be that `hX` is not an instance. See my recent suggestion for a possible fix.

tb65536 commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,487 @@
+      split,
+
+      { -- stabilizer G x ≠ ⊤ :
+        obtain ⟨ y, hy ⟩ := @exists_ne _ hX x,
```
```suggestion
        obtain ⟨y, hy⟩ := exists_ne x,
```

tb65536 commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,487 @@
+theorem is_primitive_of_two_trans
+  : (∀ (x₁ x₂ y₁ y₂ : X) (hx: x₁ ≠ x₂) (hy: y₁ ≠ y₂), ∃ g: G,
+  g • x₁  = y₁ ∧ g • x₂ = y₂ ) →
+  nontrivial X → is_primitive G X :=
```
I haven't tried this on my computer, but I suspect this might work. It marks `nontrivial X` as an instance, so that Lean knows to look for it.
```suggestion
theorem is_primitive_of_two_trans (h2X: ∀ (x₁ x₂ y₁ y₂ : X) (hx: x₁ ≠ x₂) (hy: y₁ ≠ y₂),
  ∃ g: G, g • x₁ = y₁ ∧ g • x₂ = y₂) [hX : nontrivial X] : is_primitive G X :=
```

AntoineChambert-Loir replied:
That works, thanks !

tb65536 replied:
Great. Can you push the change then?

Commit:
style: remove some @

Commit:
style: adjust comment for docstring

Commit:
style: adjust comment for docstring

tb65536 commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,486 @@
+    rw ← mul_hom.map_mul φ b' a',
+    apply φ.congr_arg,
+    refine is_commutative.cases_on is_comm _, intro,
+    exact comm a' b',
```
```suggestion
  refine λ hφ hG, ⟨λ a b, _⟩,
  obtain ⟨a, rfl⟩ := hφ a,
  obtain ⟨b, rfl⟩ := hφ b,
  rw [←φ.map_mul, ←φ.map_mul, hG.comm],
```

tb65536 commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,486 @@
+
+variables (G X)
+structure is_primitive
+extends is_transitive G X : Prop :=
```
Why not just use `is_pretransitive`? The only serious issue arises in the proof of `Iwasawa_mk`, but there you can deduce `nonempty X` from `mul_action.fixed_points N X ≠ ⊤`.
```suggestion
extends is_pretransitive G X : Prop :=
```

AntoineChambert-Loir replied:
The name of the structure should reflect its mathematical name.
As you notice, the Iwasawa lemma uses two properties of `is_primitive`, `is_transitive` and `has_maximal_stabilizers`, while the third one, as you indicate, could be derived from the other hypothesis. 
Does it make sense to define `is_preprimitive` which would just remove the `is_nonempty` field?

tb65536 replied:
Either `is_primitive` or `is_preprimitive` is fine with me.

tb65536 reviewed (commented):
In addition to these comments, can you also refactor the PR to comply with https://leanprover-community.github.io/contribute/style.html

In particular, 2 spaces to indent, spaces after binders, hypotheses before colon, single tactic per line, braces should not be alone on a line.

Commit:
feat(src/group_theory/specific_groups/alternating.lean): prove that the alternating group is perfect

Prove what is said: the alternating group on α,
with 5 ≤ fintype.card α, is perfect, aka equal to its commutator subgroup.
Prove two intermediate lemmas, that 3-cycles are commutators
and are contained in the commutator subgroup

Prove two ugly lemmas that should be (and certainly will) golfed by people
more competent than me.

eric-wieser commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,486 @@
+lemma surj_to_comm {G H : Type*} [has_mul G] [has_mul H] (φ: mul_hom G H) :
+   function.surjective φ → is_commutative G (*) → is_commutative H (*) :=
+begin
+  intros is_surj is_comm,
```
If your proof starts with `intros` then you should have put something before the colon:
```suggestion
lemma surj_to_comm {G H : Type*} [has_mul G] [has_mul H] (φ : mul_hom G H)
  (is_surj : function.surjective φ) (is_comm : is_commutative G (*)) : is_commutative H (*) :=
begin
```

AntoineChambert-Loir replied:
Oh yes ! I learned recently about the style convention for universal assumptions.

eric-wieser commented on src/group_theory/specific_groups/alternating.lean:
```diff
@@ -125,8 +137,88 @@ lemma is_three_cycle_is_conj (h5 : 5 ≤ fintype.card α)
+  }
+end
+
+private lemma stupid_lemma {A : Type*} [group A] (B : subgroup A) :
```
This should be called `subgroup.comap_subtype_self` to match `submodule.comap_subtype_self`,  and should go in `subgroup/basic.lean`.

Commit:
style(src/group_theory/iwasawa.lean): put some universal quantifiers into hypotheses

pechersky commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,486 @@
+
+namespace subgroup
+/-- An subgroup is maximal if it is maximal in the collection of proper subgroups. -/
+class is_maximal (K : subgroup G) : Prop :=
```
You make this a class, yet you never have `[is_maximal K]` anywhere.

AntoineChambert-Loir replied:
I just copied what was done for maximal ideals.

awainverse replied:
I definitely agree with this being a class to be consistent - I imagine that instances will accumulate as people do more group theory.

eric-wieser commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,483 @@
+--/
+lemma surj_to_comm {G H : Type*} [has_mul G] [has_mul H] (φ: mul_hom G H)
+   (is_surj : function.surjective φ) (is_comm : is_commutative G (*)) :
+    is_commutative H (*) :=
```
mathlib indent style, and a better name:
```suggestion
lemma function.surjective.mul_commutative {G H : Type*} [has_mul G] [has_mul H] (φ : mul_hom G H)
  (is_surj : function.surjective φ) (is_comm : is_commutative G (*)) :
  is_commutative H (*) :=
```

eric-wieser replied:
Also, I recommend using `commutative ((*) → G → G)` instead of `is_commutative G (*)`, which will eliminate the `is_commutative.mk` and `is_commutative.cases_on` below.

AntoineChambert-Loir replied:
Do you mean : `commutative ((*) : G → G → G)` ?

AntoineChambert-Loir replied:
(In any case, that worked as you indicated with this modification…)

eric-wieser replied:
Yes I did, whoops!

AntoineChambert-Loir replied:
I overlooked other implications of this modification, for the moment, the file does not compile.

AntoineChambert-Loir replied:
There are a lot of other `is_commutative` to rewrite and I wonder whether it is not clearer to refer to the structural properties of objects (groups, etc.) rather than going back to their internal law.

Is there a way to shorten : `commutative ((*) : (quotient_group.quotient N) → (quotient_group.quotient N) → (quotient_group.quotient N))` ?

eric-wieser replied:
Does `commutative ((*) : _ → _ → quotient_group.quotient N)` work?

AntoineChambert-Loir replied:
No, but you certainly know how to recover, if G is a group/monoid/... its `mul` law, or any of its types that are part of its structure. (What if there are many of them?)

eric-wieser replied:
I don't understand the question

AntoineChambert-Loir replied:
I had hoped that something like `(quotient_group.quotient N).mul` would give me the `mul` part of the group (monoid) structure of `quotient_group.quotient N`. But that didn't work like that.

eric-wieser commented on src/group_theory/iwasawa.lean:
```diff
@@ -0,0 +1,483 @@
+  intros a b,
+    obtain ⟨a', ha'⟩ := is_surj a, obtain ⟨b', hb'⟩ := is_surj b,
+    rw ← ha', rw ← hb',
+    let z := ⇑φ, let z₂ := φ.to_fun, have : z = z₂  , by refl,
```
```suggestion
    obtain ⟨a', rfl⟩ := is_surj a,
    obtain ⟨b', rfl⟩ := is_surj b,
```
you never used this `have` or `let`.

Commit:
style(src/group_theory/specific_groups/alternating.lean): remove ugly lemmas

Shorten the proof of `alternating_group_is_perfect` using a simplification
by Johan Commelin

Commit:
the good files, finally

Commit:
style: implementing suggestions of Eric Wieser

Commit:
additional corrections which were overlooked in the previous commit

Commit:
style: shorten a line (>100 chars)

Commit:
blocks and primitive actions
